File: pymab/policies/epsilon_greedy.py
"""Epsilon-greedy bandit policy."""

from __future__ import annotations

import numpy as np

from pymab.policies.greedy import GreedyPolicy
from pymab.policies.policy import choose_argmax, validate_probability


class EpsilonGreedyPolicy(GreedyPolicy):
    """Explore uniformly with probability epsilon, otherwise exploit."""

    def __init__(
        self,
        *,
        n_arms: int | None = None,
        n_bandits: int | None = None,
        initial_value: float | None = None,
        optimistic_initialization: float = 0.0,
        epsilon: float = 0.1,
        **kwargs: object,
    ) -> None:
        validate_probability(epsilon, name="epsilon")
        super().__init__(
            n_arms=n_arms,
            n_bandits=n_bandits,
            initial_value=initial_value,
            optimistic_initialization=optimistic_initialization,
            **kwargs,
        )
        self.epsilon = float(epsilon)

    def select_action(self, *, rng: np.random.Generator) -> int:
        if rng.random() < self.epsilon:
            return int(rng.integers(self.n_arms))
        return choose_argmax(self.estimates, rng)

    def __repr__(self) -> str:
        return (
            "EpsilonGreedyPolicy("
            f"epsilon={self.epsilon}, initial_value={self.initial_value})"
        )


class DecayingEpsilonGreedyPolicy(GreedyPolicy):
    """Epsilon-greedy policy with a monotone exploration schedule.

    The schedule is hyperbolic by default:
    ``epsilon_t = min_epsilon + (initial_epsilon - min_epsilon) / (1 + decay_rate * t)``.
    This gives users an intuitive policy that explores heavily early and settles
    into exploitation without fully disabling exploration unless ``min_epsilon``
    is set to zero.
    """

    def __init__(
        self,
        *,
        n_arms: int | None = None,
        n_bandits: int | None = None,
        initial_value: float | None = None,
        optimistic_initialization: float = 0.0,
        initial_epsilon: float = 1.0,
        min_epsilon: float = 0.01,
        decay_rate: float = 0.01,
        **kwargs: object,
    ) -> None:
        validate_probability(initial_epsilon, name="initial_epsilon")
        validate_probability(min_epsilon, name="min_epsilon")
        if min_epsilon > initial_epsilon:
            raise ValueError("min_epsilon must be <= initial_epsilon")
        if decay_rate < 0:
            raise ValueError("decay_rate must be non-negative")
        super().__init__(
            n_arms=n_arms,
            n_bandits=n_bandits,
            initial_value=initial_value,
            optimistic_initialization=optimistic_initialization,
            **kwargs,
        )
        self.initial_epsilon = float(initial_epsilon)
        self.min_epsilon = float(min_epsilon)
        self.decay_rate = float(decay_rate)

    @property
    def epsilon(self) -> float:
        """Current exploration probability."""

        decayed = self.min_epsilon + (
            (self.initial_epsilon - self.min_epsilon)
            / (1.0 + self.decay_rate * self.step)
        )
        return float(max(self.min_epsilon, decayed))

    def select_action(self, *, rng: np.random.Generator) -> int:
        if rng.random() < self.epsilon:
            return int(rng.integers(self.n_arms))
        return choose_argmax(self.estimates, rng)

    def __repr__(self) -> str:
        return (
            "DecayingEpsilonGreedyPolicy("
            f"initial_epsilon={self.initial_epsilon}, "
            f"min_epsilon={self.min_epsilon}, decay_rate={self.decay_rate})"
        )

File: crates/pymab-core/src/policy/epsilon_greedy.rs
//! Fixed and decaying epsilon-greedy policies.

use std::mem::size_of;

use rand::Rng;

use super::action_value::{choose_argmax, ActionValueState};
use super::Policy;
use crate::error::{PyMabError, Result};
use crate::rng::NativeRng;
use crate::types::{ActionIndex, PolicyCapabilities, PolicyObjective, ALL_REWARD_DOMAINS};
use crate::validation::probability;

const CAPABILITIES: PolicyCapabilities =
    PolicyCapabilities::new(false, ALL_REWARD_DOMAINS, PolicyObjective::CumulativeReward);

/// Explore uniformly with fixed probability epsilon and otherwise act greedily.
#[derive(Clone, Debug, PartialEq)]
pub struct EpsilonGreedyPolicy {
    epsilon: f64,
    state: ActionValueState,
}

impl EpsilonGreedyPolicy {
    /// Construct a fixed epsilon-greedy policy.
    pub fn new(n_arms: usize, initial_value: f64, epsilon: f64) -> Result<Self> {
        Ok(Self {
            epsilon: probability("epsilon", epsilon)?,
            state: ActionValueState::new(n_arms, initial_value)?,
        })
    }

    /// Return the exploration probability.
    #[must_use]
    pub const fn epsilon(&self) -> f64 {
        self.epsilon
    }
}

impl Policy for EpsilonGreedyPolicy {
    type State = ActionValueState;

    fn capabilities(&self) -> PolicyCapabilities {
        CAPABILITIES
    }

    fn n_arms(&self) -> usize {
        self.state.n_arms()
    }

    fn select_action(&mut self, rng: &mut NativeRng) -> Result<ActionIndex> {
        if rng.random::<f64>() < self.epsilon {
            ActionIndex::new(rng.random_range(0..self.n_arms()), self.n_arms())
        } else {
            choose_argmax(self.state.estimates(), rng)
        }
    }

    fn update(&mut self, action: ActionIndex, reward: f64) -> Result<()> {
        self.state.update(action, reward)
    }

    fn recommend_action(&self) -> Result<ActionIndex> {
        self.state.recommendation()
    }

    fn reset(&mut self) {
        self.state.reset();
    }

    fn state(&self) -> &Self::State {
        &self.state
    }

    fn estimated_state_bytes(&self) -> usize {
        size_of::<Self>() + self.state.estimated_heap_bytes()
    }
}

/// Epsilon-greedy policy with a hyperbolically decaying exploration schedule.
#[derive(Clone, Debug, PartialEq)]
pub struct DecayingEpsilonGreedyPolicy {
    initial_epsilon: f64,
    min_epsilon: f64,
    decay_rate: f64,
    state: ActionValueState,
}

impl DecayingEpsilonGreedyPolicy {
    /// Construct a decaying epsilon-greedy policy.
    pub fn new(
        n_arms: usize,
        initial_value: f64,
        initial_epsilon: f64,
        min_epsilon: f64,
        decay_rate: f64,
    ) -> Result<Self> {
        let initial_epsilon = probability("initial_epsilon", initial_epsilon)?;
        let min_epsilon = probability("min_epsilon", min_epsilon)?;
        if min_epsilon > initial_epsilon {
            return Err(PyMabError::configuration(
                "min_epsilon",
                "must be less than or equal to initial_epsilon",
            ));
        }
        if !decay_rate.is_finite() || decay_rate < 0.0 {
            return Err(PyMabError::configuration(
                "decay_rate",
                "must be finite and non-negative",
            ));
        }
        Ok(Self {
            initial_epsilon,
            min_epsilon,
            decay_rate,
            state: ActionValueState::new(n_arms, initial_value)?,
        })
    }

    /// Return the current exploration probability.
    #[must_use]
    pub fn epsilon(&self) -> f64 {
        let decayed = self.min_epsilon
            + (self.initial_epsilon - self.min_epsilon)
                / (1.0 + self.decay_rate * self.state.step() as f64);
        self.min_epsilon.max(decayed)
    }
}

impl Policy for DecayingEpsilonGreedyPolicy {
    type State = ActionValueState;

    fn capabilities(&self) -> PolicyCapabilities {
        CAPABILITIES
    }

    fn n_arms(&self) -> usize {
        self.state.n_arms()
    }

    fn select_action(&mut self, rng: &mut NativeRng) -> Result<ActionIndex> {
        if rng.random::<f64>() < self.epsilon() {
            ActionIndex::new(rng.random_range(0..self.n_arms()), self.n_arms())
        } else {
            choose_argmax(self.state.estimates(), rng)
        }
    }

    fn update(&mut self, action: ActionIndex, reward: f64) -> Result<()> {
        self.state.update(action, reward)
    }

    fn recommend_action(&self) -> Result<ActionIndex> {
        self.state.recommendation()
    }

    fn reset(&mut self) {
        self.state.reset();
    }

    fn state(&self) -> &Self::State {
        &self.state
    }

    fn estimated_state_bytes(&self) -> usize {
        size_of::<Self>() + self.state.estimated_heap_bytes()
    }
}
